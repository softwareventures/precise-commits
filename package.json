{
  "name": "@softwareventures/precise-commits",
  "version": "0.0.0-development",
  "description": "precise-commits",
  "author": "James Henry <james@henry.sc>",
  "contributors": [
    "Daniel Cassidy <mail@danielcassidy.me.uk>"
  ],
  "homepage": "https://github.com/softwareventures/precise-commits",
  "bugs": "https://github.com/softwareventures/precise-commits/issues",
  "repository": "github:softwareventures/precise-commits",
  "license": "MIT",
  "scripts": {
    "pre-commit": "node lib/cli.js",
    "cz": "git-cz",
    "fix": "tsc --noEmit && prettier --write .",
    "lint": "tsc --noEmit && prettier --check .",
    "prepare": "husky install && rimraf lib && tsc",
    "semantic-release": "semantic-release",
    "test": "yarn && jest"
  },
  "main": "lib/index.js",
  "bin": "lib/cli.js",
  "files": [
    "lib"
  ],
  "engines": {
    "node": "^14 || ^16 || >=18"
  },
  "dependencies": {
    "@softwareventures/array": "6.0.3",
    "@softwareventures/nullable": "3.1.0",
    "execa": "5.1.1",
    "find-up": "5.0.0",
    "glob": "8.0.3",
    "ignore": "5.2.0",
    "mri": "1.2.0",
    "ora": "5.4.1",
    "tslib": "2.4.1",
    "unknown": "0.2.5"
  },
  "peerDependencies": {
    "prettier": ">=1.8.0"
  },
  "devDependencies": {
    "@softwareventures/prettier-config": "4.0.0",
    "@softwareventures/semantic-release-config": "3.1.0",
    "@softwareventures/tsconfig": "7.0.0",
    "@types/find-up": "2.1.1",
    "@types/glob": "8.0.0",
    "@types/jest": "29.2.4",
    "@types/mkdirp": "1.0.2",
    "@types/node": "16.18.7",
    "@types/prettier": "2.7.1",
    "cz-conventional-changelog": "3.3.0",
    "husky": "8.0.2",
    "jest": "29.3.1",
    "mkdirp": "1.0.4",
    "prettier": "2.8.1",
    "rimraf": "3.0.2",
    "semantic-release": "17.4.7",
    "ts-jest": "29.0.3",
    "typescript": "4.9.4"
  },
  "prettier": "@softwareventures/prettier-config",
  "config": {
    "commitizen": {
      "path": "./node_modules/cz-conventional-changelog"
    }
  },
  "release": {
    "extends": "@softwareventures/semantic-release-config"
  },
  "jest": {
    "globalSetup": "<rootDir>/test/global-setup.js",
    "globalTeardown": "<rootDir>/test/global-teardown.js",
    "watchPathIgnorePatterns": [
      "<rootDir>/tmp"
    ],
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testRegex": ".*\\.spec\\.ts$",
    "moduleFileExtensions": [
      "ts",
      "js",
      "json"
    ]
  },
  "publishConfig": {
    "access": "public"
  }
}
