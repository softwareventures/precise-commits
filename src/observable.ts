import {Observable, Subscription} from "rxjs";

export interface Sink<T> {
    subscribed: () => boolean;
    emit: (value: T) => void;
    emitAll: (values: Observable<T>) => Promise<void>;
}

export class EmitAfterUnsubscribeError extends Error {
    public constructor() {
        super("Emitted a value after subscriber unsubscribed");
    }
}

export function observeAsync<T>(action: (sink: Sink<T>) => Promise<void>): Observable<T> {
    return new Observable<T>(subscriber => {
        let subscribed = true;
        const subscription = new Subscription(() => {
            subscribed = false;
        });

        const emit = (value: T): void => {
            if (subscribed) {
                void subscriber.next(value);
            } else {
                throw new EmitAfterUnsubscribeError();
            }
        };

        void action({
            subscribed: () => subscribed,
            emit,
            emitAll: async values =>
                new Promise((resolve, reject) => {
                    const innerSubscription = values.subscribe({
                        next: value => void emit(value),
                        error: (error: unknown) => {
                            subscriber.error(error);
                            reject(error);
                        },
                        complete: () => {
                            resolve();
                            subscription.remove(innerSubscription);
                        }
                    });
                    subscription.add(innerSubscription);
                })
        }).then(
            () => {
                subscriber.complete();
                subscription.unsubscribe();
            },
            (error: unknown) => {
                subscriber.error(error);
                subscriber.complete();
                subscription.unsubscribe();
            }
        );

        return subscription;
    });
}
