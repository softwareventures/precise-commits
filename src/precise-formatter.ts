import type {CharacterRange} from "./utils";

/**
 * Any registered precise-formatters must implement this interface
 */
export interface PreciseFormatter<FormatterConfig> {
    /**
     * Resolve the formatter config which is relevant for the given file path.
     */
    resolveConfig: (modifiedFilePath: string) => FormatterConfig | null;
    /**
     * Return true if the file contents are already formatted in accordance
     * with the given config.
     */
    isAlreadyFormatted: (fileContents: string, formatterConfig: FormatterConfig | null) => boolean;
    /**
     * Check the formatting of the file contents, using the given config
     * and character range info.
     */
    checkFormattingOfRanges: (
        filePath: string,
        fileContents: string,
        config: FormatterConfig | null,
        characterRanges: CharacterRange[]
    ) => boolean;
    /**
     * Run the formatter on the file contents, using the given config and
     * character range info.
     */
    formatRanges: (
        filePath: string,
        fileContents: string,
        config: FormatterConfig | null,
        characterRanges: CharacterRange[]
    ) => string;
    /**
     * Function which creates a callback function which will be
     * used as a predicate when filtering the modified files returned
     * from the git analysis.
     */
    generateIgnoreFilePredicate: (workingDirectory: string) => (filename: string) => boolean;
    /**
     * Return true if the given filename has a file extension which is
     * supported by the formatter.
     */
    hasSupportedFileExtension: (filename: string) => boolean;
}
