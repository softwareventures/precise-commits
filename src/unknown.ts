export function assertInstanceOf<TClass, TArgs extends unknown[]>(
    value: unknown,
    type: new (...args: TArgs) => TClass
): TClass {
    if (value instanceof type) {
        return value;
    } else {
        throw new TypeError(`Expected ${type.name}`);
    }
}
