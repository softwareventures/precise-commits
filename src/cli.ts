#!/usr/bin/env node

import {normalize} from "path";
import ora = require("ora");
import mri = require("mri");
import glob = require("glob");
import {hasProperty} from "unknown";
import {concatMap, isArray} from "@softwareventures/array";
import {pairwise} from "rxjs";
import {main} from ".";

const libraryName = "precise-commits";
const config = mri<unknown>(process.argv.slice(2));

/**
 * If the user provided one or more glob patterns to match against, ensure that there are
 * applicable files available
 */
let filesWhitelist: string[] | null = null;
if (hasProperty(config, "whitelist")) {
    if (isArray(config.whitelist)) {
        filesWhitelist = concatMap(config.whitelist, entry => glob.sync(String(entry)));
    } else {
        filesWhitelist = glob.sync(String(config.whitelist));
    }
    if ((filesWhitelist?.length ?? 0) === 0) {
        console.error(
            `Error: No files match the glob pattern(s) you provided for --whitelist -> "${String(
                config.whitelist
            )}"`
        );
        process.exit(1);
    }
    filesWhitelist = filesWhitelist.map(normalize);
}

/**
 * If the user specifies at least one SHA, perform some validation and
 * apply some defaults
 */
let base: string | null = null;
let head: string | null = null;
if (hasProperty(config, "base") || hasProperty(config, "head")) {
    if (!hasProperty(config, "base") || config.base == null) {
        console.error(
            `Error: When giving a value of --head, you must also give a value for --base`
        );
        process.exit(1);
    }

    if (hasProperty(config, "head") && config.head != null) {
        // FIXME: Tech debt
        // eslint-disable-next-line @typescript-eslint/no-base-to-string
        head = String(config.head);
    } else {
        /**
         * If the user only specified `--base`, set the value of `--head` to be "HEAD"
         */
        head = "HEAD";
    }

    // FIXME: Tech debt
    // eslint-disable-next-line @typescript-eslint/no-base-to-string
    base = String(config.base);
}

const checkOnly = hasProperty(config, "check-only") ? Boolean(config["check-only"]) : false;
const formatter = hasProperty(config, "formatter")
    ? // FIXME: Tech debt
      // eslint-disable-next-line @typescript-eslint/no-base-to-string
      String(config.formatter ?? "prettier")
    : "prettier";

const options = {
    checkOnly,
    filesWhitelist,
    base,
    head,
    formatter
};

const spinner = ora(`Running ${libraryName}`);

let shouldErrorOut = false;

main(process.cwd(), options)
    .pipe(pairwise())
    .subscribe({
        next: ([previous, current]) => {
            if (current.state === "Running") {
                if (previous.state === "Initializing") {
                    spinner.start(`Running ${libraryName}`);
                }

                if (current.files.length > 0) {
                    spinner.text = `${libraryName}: ${current.files.length} modified file(s) found`;
                }

                current.files.forEach((fileState, index) => {
                    const previousFileState =
                        previous.state === "Running" ? previous.files[index] : undefined;

                    if (previousFileState == null) {
                        spinner.info(`Processing file: ${fileState.filename}`);
                    }

                    if (previousFileState?.status !== fileState.status) {
                        if (fileState.status === "Updated") {
                            spinner.succeed(`Updated formatting in: ${fileState.filename}`);
                        } else if (fileState.status === "NotUpdated") {
                            spinner.succeed(
                                `No formatting changes required in: ${fileState.filename}`
                            );
                        } else if (fileState.status === "InvalidFormatting") {
                            // If --check-only is passed as a CLI argument, the script will error out.
                            if (options.checkOnly) {
                                shouldErrorOut = true;
                            }

                            spinner.fail(`Invalid formatting detected in: ${fileState.filename}`);
                        }
                    }
                });
            } else if (current.state === "Finished") {
                if (current.fileCount === 0) {
                    spinner.info("No matching modified files detected.");
                    spinner.info(
                        "If you feel that one or more files should be showing up here, be sure to first check what file extensions prettier supports, and whether or not you have included those files in a .prettierignore file"
                    );
                    spinner.stop();
                } else if (options.checkOnly) {
                    spinner.succeed("Checks complete ðŸŽ‰");
                } else {
                    spinner.succeed("Formatting complete ðŸŽ‰");
                }
                return process.exit(shouldErrorOut ? 1 : 0);
            }
        },
        error: (error: unknown) => {
            spinner.fail("An Error occurred");
            console.error("\n");
            console.error(error);
            console.error("\n");
            spinner.stop();
            return process.exit(1);
        }
    });
