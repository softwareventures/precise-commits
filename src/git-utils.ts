import {sync as findUpSync} from "find-up";
import {dirname} from "path";

import {runCommandSync} from "./utils";
import {hasProperty} from "unknown";

interface DiffIndexFile {
    diffFilterChar: string;
    filename: string;
}

export function resolveNearestGitDirectoryParent(workingDirectory: string) {
    const gitDirectoryPath = findUpSync(".git", {cwd: workingDirectory, type: "directory"});
    if (!gitDirectoryPath) {
        throw new Error("No .git directory found");
    }
    return dirname(gitDirectoryPath);
}

export function getDiffForFile(
    gitDirectoryParent: string,
    fullPath: string,
    base: string | null,
    head: string | null
): string {
    if (base && head) {
        return runCommandSync(
            "git",
            ["diff", "--unified=0", base, head, fullPath],
            gitDirectoryParent
        ).stdout;
    }
    return runCommandSync("git", ["diff", "--unified=0", "--cached", fullPath], gitDirectoryParent)
        .stdout;
}

/**
 * Output of `git diff-index --help`:
 *
 * --diff-filter=[(A|C|D|M|R|T|U|X|B)...[*]]
      Select only files that are Added (A), Copied (C), Deleted (D), Modified (M), Renamed (R), have
      their type (i.e. regular file, symlink, submodule, ...) changed (T), are Unmerged (U), are
      Unknown (X), or have had their pairing Broken (B). Any combination of the filter characters
      (including none) can be used. When * (All-or-none) is added to the combination, all paths are
      selected if there is any file that matches other criteria in the comparison; if there is no
      file that matches other criteria, nothing is selected.

      Also, these upper-case letters can be downcased to exclude. E.g.  --diff-filter=ad excludes
      added and deleted paths.
 *
 */
/**
 * const DIFF_INDEX_FILTER = 'ACDMRTUXB';
 * NOTE: We are only explicitly testing "Modified" and "Added" files for now...
 */
const DIFF_INDEX_FILTER = "AM";
const SPECIAL_EMPTY_TREE_COMMIT_HASH = "4b825dc642cb6eb9a060e54bf8d69288fbee4904";

export function getModifiedFilenames(
    gitDirectoryParent: string,
    base: string | null,
    head: string | null
): string[] {
    let diffIndexOutput: string;
    if (base && head) {
        /**
         * We are grabbing the files between the two given commit SHAs
         */
        diffIndexOutput = runCommandSync(
            "git",
            ["diff", "--name-status", `--diff-filter=${DIFF_INDEX_FILTER}`, base, head],
            gitDirectoryParent
        ).stdout;
    } else {
        /**
         * No commit SHAs given, we assume we are attempting to evaluate staged files,
         * and so need to determine if there is HEAD SHA available.
         */
        let head: string = "";
        try {
            head = runCommandSync(
                "git",
                ["rev-parse", "--verify", "HEAD"],
                gitDirectoryParent
            ).stdout.replace("\n", "");
        } catch (err) {
            /**
             * If there has never been a commit before, there will be no HEAD to compare
             * to. Use the special empty tree hash value instead:
             * https://stackoverflow.com/questions/9765453/is-gits-semi-secret-empty-tree-object-reliable-and-why-is-there-not-a-symbolic
             */
            if (
                hasProperty(err, "message") &&
                typeof err.message === "string" &&
                err.message.includes("fatal: Needed a single revision")
            ) {
                head = SPECIAL_EMPTY_TREE_COMMIT_HASH;
            } else {
                throw err;
            }
        }
        diffIndexOutput = runCommandSync(
            "git",
            ["diff-index", "--cached", "--name-status", `--diff-filter=${DIFF_INDEX_FILTER}`, head],
            gitDirectoryParent
        ).stdout;
    }
    const allFiles = parseDiffIndexOutput(diffIndexOutput);
    return allFiles.map(r => r.filename);
}

function parseDiffIndexOutput(stdout: string): DiffIndexFile[] {
    const lines = stdout.split("\n");
    return lines.filter(Boolean).map(line => {
        const parts = line.split("\t");
        return {
            filename: parts[1],
            diffFilterChar: parts[0]
        };
    });
}
