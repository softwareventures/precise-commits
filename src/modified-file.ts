import {readFileSync, writeFileSync} from "fs";
import {relative, sep, posix} from "path";
import {notNull} from "@softwareventures/nullable";
import execa = require("execa");
import type {CharacterRange} from "./utils";
import {
    calculateCharacterRangesFromLineChanges,
    extractLineChangeData
} from "./utils";
import {getDiffForFile, index, workingTree} from "./git-utils";
import type {PreciseFormatter} from "./precise-formatter";
import {assertInstanceOf} from "./unknown";

export interface ModifiedFileConfig {
    fullPath: string;
    gitDirectoryParent: string;
    base: string | null;
    head: string | typeof index | typeof workingTree;
    selectedFormatter: PreciseFormatter<any>;
}

export class ModifiedFile {
    private readonly fullPath: string;
    private readonly pathInGit: string;
    /**
     * An optional commit SHA pair which will be used to inform how the git
     * commands are run. E.g. `git diff`
     */
    private readonly base: string | null;
    private readonly head: string | typeof index | typeof workingTree;
    /**
     * The chosen formatter to be run on the modified file.
     */
    private readonly selectedFormatter: PreciseFormatter<any>;
    /**
     * The parent directory of the relevant .git directory that was resolved
     * for the modified file.
     */
    private readonly gitDirectoryParent: string;
    /**
     * The contents of the file in their current state on the user's file
     * system
     */
    private readonly fileContents: string;
    /**
     * The final file contents, after we've run the formatter
     */
    private formattedFileContents: string | null = null;
    /**
     * The resolved formatter config which applies to this file
     */
    private readonly formatterConfig: object | null;
    /**
     * The calculated character ranges which have been modified
     * within this file
     */
    private modifiedCharacterRanges: CharacterRange[] = [];

    constructor({fullPath, gitDirectoryParent, base, head, selectedFormatter}: ModifiedFileConfig) {
        this.fullPath = fullPath;
        this.pathInGit = relative(gitDirectoryParent, fullPath).split(sep).join(posix.sep);
        this.gitDirectoryParent = gitDirectoryParent;
        this.base = base;
        this.head = head;
        this.selectedFormatter = selectedFormatter;
        this.fileContents =
            head === workingTree
                ? readFileSync(this.fullPath, "utf8")
                : head === index
                ? execa.sync("git", ["show", `:0:${this.pathInGit}`], {
                      cwd: gitDirectoryParent,
                      stripFinalNewline: false
                  }).stdout
                : execa.sync("git", ["show", `${head}:${this.pathInGit}`], {
                      cwd: gitDirectoryParent,
                      stripFinalNewline: false
                  }).stdout;
        this.formatterConfig = this.selectedFormatter.resolveConfig(this.fullPath);
    }

    /**
     * Return true if the whole file has already been formatted appropriately based on
     * the resolved formatter config. We can use this as a check to skip unnecessary work.
     */
    isAlreadyFormatted(): boolean {
        return this.selectedFormatter.isAlreadyFormatted(this.fileContents, this.formatterConfig);
    }

    /**
     * Run the formatters check mode on the given ranges and return true if they are all
     * already formatted appropriately based on the resolved formatter config.
     */
    hasValidFormattingForCharacterRanges(): boolean {
        return this.selectedFormatter.checkFormattingOfRanges(
            this.fullPath,
            this.fileContents,
            this.formatterConfig,
            this.modifiedCharacterRanges
        );
    }

    /**
     * Run the formatter on the file contents and store the result
     */
    formatCharacterRangesWithinContents(): void {
        this.formattedFileContents = this.selectedFormatter.formatRanges(
            this.fullPath,
            this.fileContents,
            this.formatterConfig,
            this.modifiedCharacterRanges
        );
    }

    /**
     * Return true if the formatted file contents are different to
     * what was originally resolved from disk.
     */
    shouldContentsBeUpdatedOnDisk(): boolean {
        return this.fileContents !== this.formattedFileContents;
    }

    /**
     * Write the updated file contents back to disk.
     */
    updateFileOnDisk(): void {
        if (this.head === index) {
            const hash = execa.sync(
                "git",
                ["hash-object", "-w", "--path", this.fullPath, "--stdin"],
                {
                    cwd: this.gitDirectoryParent,
                    input: notNull(this.formattedFileContents)
                }
            ).stdout;
            const mode = execa
                .sync("git", ["ls-files", "--stage", "--", this.fullPath], {
                    cwd: this.gitDirectoryParent
                })
                .stdout.split(" ")?.[0];
            if (mode == null) {
                throw new Error("Can't find file in git index");
            }
            execa.sync(
                "git",
                [
                    "update-index",
                    "--add",
                    "--replace",
                    "--cacheinfo",
                    `100644,${hash},${this.pathInGit}`
                ],
                {cwd: this.gitDirectoryParent}
            );
        } else {
            writeFileSync(this.fullPath, notNull(this.formattedFileContents));
        }
    }

    /**
     * We handle errors locally within this method to allow for
     * more granular feedback within the main() function of the
     * library.
     */
    calculateModifiedCharacterRanges(): {err: Error | null} {
        try {
            /**
             * Extract line change data from the git diff results.
             */
            const diff = getDiffForFile(
                this.gitDirectoryParent,
                this.fullPath,
                this.base,
                this.head
            );
            const lineChangeData = extractLineChangeData(diff);
            /**
             * Convert the line change data into character data.
             */
            this.modifiedCharacterRanges = calculateCharacterRangesFromLineChanges(
                lineChangeData,
                this.fileContents
            );
            return {err: null};
        } catch (err) {
            return {err: assertInstanceOf(err, Error)};
        }
    }
}
