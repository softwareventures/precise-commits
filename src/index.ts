import {isAbsolute, join, relative} from "path";

import type {Observable} from "rxjs";
import {concatAll, count, endWith, filter, from, map, mergeMap, mergeWith, of, scan} from "rxjs";
import {
    getModifiedFilenames,
    index,
    resolveNearestGitDirectoryParent,
    workingTree
} from "./git-utils";
import {generateFilesWhitelistPredicate, noLineChangeDataError} from "./utils";
import {preciseFormatterPrettier} from "./precise-formatters/prettier";
import {ModifiedFile} from "./modified-file";

export interface AdditionalOptions {
    checkOnly: boolean;
    filesWhitelist: string[] | null;
    base: string | null;
    head: string | null;
}

export type State = Initializing | Running | Finished;

export interface Initializing {
    readonly state: "Initializing";
}

export interface Running {
    readonly state: "Running";
    readonly gitSearchComplete: boolean;
    readonly files: readonly FileState[];
}

export interface FileState {
    readonly filename: string;
    readonly status: FileStatus;
}

export type FileStatus = "Processing" | "NotUpdated" | "Updated" | "InvalidFormatting";

export interface Finished {
    readonly state: "Finished";
    readonly fileCount: number;
}

/**
 * LIBRARY
 */
export function main(
    workingDirectory: string,
    additionalOptions: AdditionalOptions
): Observable<State> {
    // Merge user-given and default options.
    return of({
        ...{
            filesWhitelist: null,
            base: null,
            head: null,
            checkOnly: false,
            formatter: "prettier"
        },
        ...additionalOptions
    }).pipe(
        mergeMap(async options => {
            // Note: Will be exposed as an option if/when new formatters are added.
            if (options.formatter !== "prettier") {
                throw new Error(`The only supported value for "formatter" option is "prettier"`);
            }

            const selectedFormatter = preciseFormatterPrettier;

            // Resolve the relevant .git directory's parent directory up front, as we will need this when
            // executing various `git` commands.
            const gitDirectoryParent = await resolveNearestGitDirectoryParent(workingDirectory);

            // Find files that we should process. A file is relevant if:
            //  * The file extension is supported by the given formatter.
            //  * The file is included in the optional `filesWhitelist` array, or no whitelist is specified.
            //  * The file is not ignored as a result of any supported "ignore" mechanism of the formatter.
            const relevantFiles = from(
                getModifiedFilenames(gitDirectoryParent, options.base, options.head)
            ).pipe(
                mergeMap(array => array),
                map(path => join(gitDirectoryParent, path)),
                map(path => relative(workingDirectory, path)),
                filter(path => !isAbsolute(path)),
                filter(selectedFormatter.hasSupportedFileExtension),
                filter(generateFilesWhitelistPredicate(options.filesWhitelist)),
                filter(selectedFormatter.generateIgnoreFilePredicate(workingDirectory))
            );

            const newFileEvents = relevantFiles.pipe(
                map((filename, fileIndex) => ({event: "NewFile", fileIndex, filename} as const))
            );

            const gitSearchCompleteEvent = relevantFiles.pipe(
                count(),
                map(fileCount => ({event: "GitSearchComplete", fileCount} as const))
            );

            const fileProcessedEvents = relevantFiles.pipe(
                mergeMap(async (filename, fileIndex) => {
                    const fullPath = join(workingDirectory, filename);

                    // Read the modified file contents and resolve the relevant formatter.
                    const modifiedFile = new ModifiedFile({
                        fullPath,
                        gitDirectoryParent,
                        base: options.base,
                        head: options.head ?? index,
                        selectedFormatter
                    });

                    // To avoid unnecessary issues with 100% valid files producing issues when parts
                    // of them are reformatted in isolation, we first check the whole file to see if
                    // it is already formatted. This could also allow us to skip unnecessary git diff
                    // analysis work.
                    if (modifiedFile.isAlreadyFormatted()) {
                        return {
                            event: "FileProcessed",
                            fileIndex,
                            filename,
                            status: "NotUpdated"
                        } as const;
                    }

                    // Calculate what character ranges have been affected in the modified file.
                    // If any of the analysis threw an error for any reason, it will be returned
                    // from the method so we can handle it here.
                    const {err} = modifiedFile.calculateModifiedCharacterRanges();
                    if (err != null) {
                        if (err.message === noLineChangeDataError) {
                            return {
                                event: "FileProcessed",
                                fileIndex,
                                filename,
                                status: "NotUpdated"
                            } as const;
                        }

                        // Unexpected error
                        throw err;
                    }

                    // "CHECK ONLY MODE"
                    if (options.checkOnly) {
                        return {
                            event: "FileProcessed",
                            fileIndex,
                            filename,
                            status: modifiedFile.hasValidFormattingForCharacterRanges()
                                ? "NotUpdated"
                                : "InvalidFormatting"
                        } as const;
                    }

                    // "FORMAT MODE"
                    modifiedFile.formatCharacterRangesWithinContents();
                    if (!modifiedFile.shouldContentsBeUpdatedOnDisk()) {
                        return {
                            event: "FileProcessed",
                            fileIndex,
                            filename,
                            status: "NotUpdated"
                        } as const;
                    }

                    // If we're updating the index, we also need to update the working tree.
                    if (options.head == null) {
                        const workingTreeFile = new ModifiedFile({
                            fullPath,
                            gitDirectoryParent,
                            base: options.base,
                            head: workingTree,
                            selectedFormatter
                        });
                        const {err} = workingTreeFile.calculateModifiedCharacterRanges();
                        if (err == null) {
                            workingTreeFile.formatCharacterRangesWithinContents();
                            if (workingTreeFile.shouldContentsBeUpdatedOnDisk()) {
                                workingTreeFile.updateFileOnDisk();
                            }
                        } else if (err.message !== noLineChangeDataError) {
                            // Unexpected error
                            throw err;
                        }
                    }

                    // Write the file back to disk and report.
                    modifiedFile.updateFileOnDisk();
                    return {
                        event: "FileProcessed",
                        fileIndex,
                        filename,
                        status: "Updated"
                    } as const;
                })
            );

            return newFileEvents.pipe(mergeWith(gitSearchCompleteEvent, fileProcessedEvents));
        }),
        concatAll(),
        endWith({event: "Finished"} as const),
        scan(
            (state: State, event) => {
                if (event.event === "NewFile") {
                    if (state.state === "Running") {
                        return {
                            ...state,
                            files: [
                                ...state.files,
                                {
                                    filename: event.filename,
                                    status: "Processing"
                                }
                            ]
                        } as const;
                    } else {
                        return {
                            state: "Running",
                            gitSearchComplete: false,
                            files: [{filename: event.filename, status: "Processing"}]
                        } as const;
                    }
                } else if (event.event === "GitSearchComplete") {
                    if (state.state === "Running") {
                        return {...state, gitSearchComplete: true} as const;
                    } else {
                        return {state: "Running", gitSearchComplete: true, files: []} as const;
                    }
                } else if (event.event === "FileProcessed") {
                    if (state.state === "Running") {
                        const files = [...state.files];
                        files[event.fileIndex] = {
                            filename: event.filename,
                            status: event.status
                        };
                        return {...state, files} as const;
                    } else {
                        return {
                            state: "Running",
                            gitSearchComplete: false,
                            files: [
                                {
                                    filename: event.filename,
                                    status: event.status
                                }
                            ]
                        } as const;
                    }
                } else if (event.event === "Finished") {
                    if (state.state === "Running") {
                        return {state: "Finished", fileCount: state.files.length} as const;
                    } else {
                        return {state: "Finished", fileCount: 0} as const;
                    }
                } else {
                    return state;
                }
            },
            {state: "Initializing"}
        )
    );
}
