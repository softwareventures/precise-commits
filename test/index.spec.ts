import {readFileSync} from "fs";
import {lastValueFrom, tap} from "rxjs";
import {main} from "../src/index";
import {name as libraryName} from "../package.json";
import {TestBed, readFixtures, mergeOptionsForTmpFile} from "./test-utils";

const fixtures = readFixtures();
let testBed: TestBed;

describe(libraryName, () => {
    describe("main()", () => {
        beforeAll(() => {
            testBed = new TestBed();
        });

        fixtures.forEach(fixture => {
            it(fixture.fixtureName, async () => {
                expect.assertions(1);
                testBed.prepareFixtureInTmpDirectory(fixture);
                const tmpFile = testBed.getTmpFileForFixture(fixture);
                const options = mergeOptionsForTmpFile(
                    {checkOnly: false, filesWhitelist: null},
                    tmpFile
                );
                await lastValueFrom(main(tmpFile.directoryPath, options));
                const formatted = readFileSync(tmpFile.path, "utf8");
                expect(formatted).toMatchSnapshot();
            });
        });
    });

    describe("main() - checkOnly: true", () => {
        beforeAll(() => {
            testBed = new TestBed();
        });

        fixtures.forEach(fixture => {
            it(fixture.fixtureName, async () => {
                expect.assertions(1);
                testBed.prepareFixtureInTmpDirectory(fixture);
                const tmpFile = testBed.getTmpFileForFixture(fixture);
                const options = mergeOptionsForTmpFile(
                    {checkOnly: true, filesWhitelist: null},
                    tmpFile
                );
                await lastValueFrom(
                    main(tmpFile.directoryPath, options).pipe(
                        tap(event => {
                            if (event.event === "FinishedProcessingFile") {
                                expect(event.status).toEqual("INVALID_FORMATTING");
                            }
                        })
                    )
                );
            });
        });
    });
});
