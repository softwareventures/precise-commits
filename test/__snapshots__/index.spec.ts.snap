// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@softwareventures/precise-commits main() brand-new-json-file 1`] = `
"{
  "foo": true
}
"
`;

exports[`@softwareventures/precise-commits main() brand-new-jsx-file 1`] = `
"function HelloWorld({greeting = "hello", greeted = '"World"', silent = false, onMouseOver}) {
    if (!greeting) {
        return null;
    }

    // TODO: Don't use random in render
    let num = Math.floor(Math.random() * 1e7)
        .toString()
        .replace(/\\.\\d+/gi, "");

    return (
        <div
            className="HelloWorld"
            title={\`You are visitor number \${num}\`}
            onMouseOver={onMouseOver}
        >
            <strong>{greeting.slice(0, 1).toUpperCase() + greeting.slice(1).toLowerCase()}</strong>
            {greeting.endsWith(",") ? " " : <span style={{color: "\\grey"}}>", "</span>}
            <em>{greeted}</em>
            {silent ? "." : "!"}
        </div>
    );
}
"
`;

exports[`@softwareventures/precise-commits main() changed-line-in-closure 1`] = `
"import {Egg} from "egg";
import {waitFor} from 'wait-for'

const failOnError = error => {
    console.log("got error:", error);
    fail(error);
};

describe("salad sausage", () => {

    it("hats notepad tree", async () => {
        const subscription = repo.findAbc(111, Egg.fromText("nettle"));
        subscription.getSnapshots().subscribe(
            result => {
                console.info("got result: ", result);
                firstResult = result;
            },
            failOnError,
            () => {
                flag = true;
                console.log("finished"); // TODO: winkle ankle cricket
            }
        );
        subscription.getEffects().subscribe(effect => commandEffectsReceived.push(effect));
        return waitFor(() => {
            expect(flag).toBeTruthy();
            expect(firstResult).toBeNull();
            expect(commandEffectsReceived).toBeEmpty();
        }, {timeout: 150});
    });
});
"
`;

exports[`@softwareventures/precise-commits main() empty-javascript-file 1`] = `
"var greeting = "hello";
var audience = "world";
"
`;

exports[`@softwareventures/precise-commits main() formatting-would-affect-total-line-count 1`] = `
"  var untouchedUglyLine     =      10

function myfunc(a, b, c) {
    console.log("blah");
}

var newStatement = true;
"
`;

exports[`@softwareventures/precise-commits main() hunk-marker-in-code 1`] = `
"export function extractLineChangeData(diffData: string) {
    return diffData.match(/@@.*@@/gu);
}"
`;

exports[`@softwareventures/precise-commits main() multiple-updated-lines-javascript-file 1`] = `
"var     untouchedUglyLines1 = ''
function TouchedFunctionSignature({
  greeting = "Hello",
  greeted = '"Mars"',
  silent = false
}) {}
      const newlyAddedUglyLine = true;"
`;

exports[`@softwareventures/precise-commits main() single-updated-line-typescript 1`] = `
"const a = true
const b = true;"
`;

exports[`@softwareventures/precise-commits main() updated-class-body 1`] = `
"class Hatstand {
    constructor(hats) {
        this.hats = hats ?? [];
    }
}
"
`;
